src/FixedGuns.cpp

```diff
@@ -2,228 +2,438 @@
 // Licensed under the terms of the GPL v3. See licenses/GPL-3.txt
 
 #include "FixedGuns.h"
+
 #include "Beam.h"
-#include "DynamicBody.h"
+#include "Body.h"
 #include "GameSaveError.h"
 #include "Projectile.h"
 #include "StringF.h"
+#include "scenegraph/Model.h"
 #include "scenegraph/MatrixTransform.h"
+#include "JsonUtils.h"
 
-FixedGuns::FixedGuns()
+FixedGuns::FixedGuns(Body* b)
 {
+	b->AddFeature(Body::FIXED_GUNS);
 }
 
 FixedGuns::~FixedGuns()
 {
 }
 
-bool FixedGuns::IsFiring()
-{
-	bool gunstate = false;
-	for (int j = 0; j < Guns::GUNMOUNT_MAX; j++)
-		gunstate |= m_is_firing[j];
-	return gunstate;
-}
-
-bool FixedGuns::IsFiring(const int num)
-{
-	return m_is_firing[num];
-}
-
-bool FixedGuns::IsBeam(const int num)
-{
-	return m_gun[num].projData.beam;
-}
-
-void FixedGuns::Init(DynamicBody *b)
-{
-	for (int i = 0; i < Guns::GUNMOUNT_MAX; i++) {
-		// Initialize structs
-		m_is_firing[i] = false;
-		m_gun[i].recharge = 0;
-		m_gun[i].temp_heat_rate = 0;
-		m_gun[i].temp_cool_rate = 0;
-		m_gun[i].projData.lifespan = 0;
-		m_gun[i].projData.damage = 0;
-		m_gun[i].projData.length = 0;
-		m_gun[i].projData.width = 0;
-		m_gun[i].projData.mining = false;
-		m_gun[i].projData.speed = 0;
-		m_gun[i].projData.color = Color::BLACK;
-		// Set there's no guns:
-		m_gun_present[i] = false;
-		m_recharge_stat[i] = 0.0;
-		m_temperature_stat[i] = 0.0;
-	};
-	b->AddFeature(DynamicBody::FIXED_GUNS);
-}
-
 void FixedGuns::SaveToJson(Json &jsonObj, Space *space)
 {
-
 	Json gunArray = Json::array(); // Create JSON array to contain gun data.
 
-	for (int i = 0; i < Guns::GUNMOUNT_MAX; i++) {
-		Json gunArrayEl = Json::object(); // Create JSON object to contain gun.
-		gunArrayEl["state"] = m_is_firing[i];
-		gunArrayEl["recharge"] = m_recharge_stat[i];
-		gunArrayEl["temperature"] = m_temperature_stat[i];
+	for (int i = 0; i < m_guns.size(); i++) {
+		Json gunArrayEl({});// = Json::object(); // Create JSON object to contain gun.
+		//Json projectileObj({})
+		gunArrayEl["state"] = m_guns[i].is_firing;
+		gunArrayEl["active"] = m_guns[i].is_active;
+		gunArrayEl["recharge"] = m_guns[i].recharge_stat;
+		gunArrayEl["temperature"] = m_guns[i].temperature_stat;
+		gunArrayEl["contemporary_barrels"] = m_guns[i].contemporary_barrels;
+		gunArrayEl["next_firing"] = m_guns[i].next_firing_barrels;
+		gunArrayEl["mount_name"] = m_mounts[m_guns[i].mount_id].name; // <- Save the name of hardpoint (mount)
+		// Save "GunData":
+		gunArrayEl["gd_name"] = m_guns[i].gun_data.gun_name;
+		gunArrayEl["gd_sound"] = m_guns[i].gun_data.sound;
+		gunArrayEl["gd_barrels"] = m_guns[i].gun_data.barrels;
+		gunArrayEl["gd_recharge"] = m_guns[i].gun_data.recharge;
+		gunArrayEl["gd_cool_rate"] = m_guns[i].gun_data.temp_cool_rate;
+		gunArrayEl["gd_heat_rate"] = m_guns[i].gun_data.temp_heat_rate;
+		// Save "ProjectileData":
```

Comment:
Probably save/load should be moved in the `ProjectileData` struct, making easier to add/remove properties